### PR TITLE
V4 uncertainty visualizations

### PR DESCRIPTION
**Pull request by ThomasGalesloot**
# Pull Request Key Information

## Reason for change

Visualizations to communicate variation between simulated "estimated" and "true" emissions will greatly assist modellers in understanding the effectiveness of a measurement strategy

## What was changed

Added 3 visualizations to communicate "uncertainty" in a measurement strategy

1. The percentage difference between "estimated" and "true" emissions
2. The relative difference between "estimated" and "true" emissions
3. The paired distributions of "estimated" and "true" emissions

## Intended Purpose

Visualizations to assist in evaluating the performance of a measurement strategy

## Level of version change required

N/A

## Testing Completed

Manual testing.

## Target Issue

N/A

## Additional Information

N/A

### CODE REVIEW

File: LDAR_Sim/src/file_processing/output_processing/output_constants.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+TRUE_AND_ESTIMATED_PAIRED_EMISSIONS_DISTRIBUTION_PLOT = (
+    "True_and_Estimated_Paired_Emissions_Distribution.png"
+)
+SUMMARY_PROGRAM_PLOTS_DIRECTORY = "program_summary_plots"
+
+
+class ESTIMATE_VS_TRUE_PLOTTING_CONSTANTS:
+    TRUE_EMISSIONS_SUFFIX = '"True" Total Emissions (Kg Methane)'
+    ESTIMATED_EMISSIONS_SUFFIX = '"Estimated" Total Emissions (Kg Methane)'
+    HISTOGRAM_BIN_WIDTH = 8.0
```
Comment (SEAJang) [inline LDAR_Sim/src/file_processing/output_processing/output_constants.py:21]: Should this be something that's hard set, or something that users could eventually edit-be able to set?

File: LDAR_Sim/src/file_processing/output_processing/output_constants.py
Comment (SEAJang) [inline LDAR_Sim/src/file_processing/output_processing/output_constants.py:1]: add in license at top of file

File: LDAR_Sim/src/file_processing/output_processing/output_utils.py [hunk 2]
```diff
@@ -83,6 +84,29 @@ class TIMESERIES_COL_ACCESSORS:
     METH_DAILY_SURVEY_TIME = "{method} Survey Time (Minutes)"
 
 
+def percent_difference(a: float, b: float) -> float:
```
Comment (SEAJang) [inline LDAR_Sim/src/file_processing/output_processing/output_utils.py:87]: Should this have error handling for div by 0?

File: LDAR_Sim/src/file_processing/output_processing/output_utils.py [hunk 3]
```diff
@@ -83,6 +84,29 @@ class TIMESERIES_COL_ACCESSORS:
...
+    return abs(a - b) / ((a + b) / 2) * 100
+
+
+def relative_difference(aquired_value: float, true_value: float) -> float:
```
Comment (SEAJang) [inline LDAR_Sim/src/file_processing/output_processing/output_utils.py:91]: Same comment as above

File: LDAR_Sim/src/ldar_sim_run.py [hunk 4]
```diff
@@ -302,3 +303,4 @@ def simulate(
             # -- Batch Report --
             print(f"...Cleaning up batch {batch_count} data")
             concat_output_data(out_dir, batch_count != 0)
+    gen_cross_program_summary_plots(out_dir, base_program)
```
Comment (SEAJang) [inline LDAR_Sim/src/ldar_sim_run.py:306]: Could add in another print here to indicate the cross program plots are being generated here.

**Review by SEAJang — COMMENTED** [review]
Minor nit-pick changes, however great job overall!

File: LDAR_Sim/src/file_processing/output_processing/multi_simulation_outputs.py [hunk 5]
```diff
@@ -118,66 +49,78 @@ def get_nth_percentile(df: pd.DataFrame, column: str, percentile: float) -> floa
     return np.percentile(df[column], percentile, method="median_unbiased")
 
 
-TS_MAPPING_TO_SUMMARY_COLS = {
-    TS_SUMMARY_COLUMNS_ACCESSORS.AVG_T_DAILY_EMIS: lambda df: get_mean_val(df, tca.EMIS),
-    TS_SUMMARY_COLUMNS_ACCESSORS.AVG_T_MIT_DAILY_EMIS: lambda df: get_mean_val(df, tca.EMIS_MIT),
-    TS_SUMMARY_COLUMNS_ACCESSORS.AVG_T_NON_MIT_DAILY_EMIS: lambda df: get_mean_val(
-        df, tca.EMIS_NON_MIT
+(TS_MAPPING_TO_SUMMARY_COLS) = {
```
[the remark this answers is not in the document]
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/multi_simulation_outputs.py:52]: Not sure - Will remove

File: LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
+import os
+from pathlib import Path
+from typing import Tuple
+import pandas as pd
+import seaborn as sns
+from matplotlib import pyplot as plt
+from matplotlib import ticker, lines
+
+from file_processing.output_processing import output_constants
+from file_processing.output_processing import output_utils
```
[the remark this answers is not in the document]
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py:10]: Done

File: LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+            == program_name,
+            "percent_diff",
+        ].tolist()
+        percent_diff_lists[program_name] = percent_diff_list
+
+    if combine_plots:
+        comb_fig: plt.Figure = plt.figure()
+        comb_ax: plt.Axes = plt.gca()
+
+    colors = sns.color_palette("husl", n_colors=len(program_names))
```
[the remark this answers is not in the document]
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py:145]: Will add a TODO and ticket in the backlog

File: LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py
[the remark this answers is not in the document]
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py:1]: Done

File: LDAR_Sim/src/file_processing/output_processing/output_constants.py (hunk 1 above)
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/output_constants.py:21]: I think this should be semi settable by users - In the sense we give them options - not let them set any number. Will add a TODO

File: LDAR_Sim/src/file_processing/output_processing/output_utils.py (hunk 2 above)
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/output_utils.py:87]: Will add in handling for the case both values are 0

File: LDAR_Sim/src/file_processing/output_processing/output_utils.py (hunk 3 above)
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/output_utils.py:91]: Will add in handling for the case the reference value is 0

File: LDAR_Sim/src/ldar_sim_run.py (hunk 4 above)
Comment (ThomasGalesloot) [reply LDAR_Sim/src/ldar_sim_run.py:306]: Will do

File: LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+    bin_range: Tuple[float, float],
+    x_label: str,
+    y_label: str,
+    legend_label: str = None,
+    bin_width: float = 0,
+    bins: int = 0,
+    x_locator: ticker.Locator = ticker.AutoLocator(),
+    y_locator: ticker.Locator = ticker.MaxNLocator(nbins="auto", integer=True, prune=None),
+):
+    if bins != 0:
```
[the remark this answers is not in the document]
Comment (ThomasGalesloot) [reply LDAR_Sim/src/file_processing/output_processing/multi_simulation_visualizations.py:57]: not sure I quite understand what you mean

**Pull request merged**